test/end2end_test.go

```diff
@@ -2349,6 +2349,60 @@ func TestNonFailFastRPCWithNoBalancerErrorOnBadCertificates(t *testing.T) {
 	}
 }
 
+type serverDispatchCred struct {
+	ready   chan int
+	rawConn net.Conn
+}
+
+func newServerDispatchCred() *serverDispatchCred {
+	return &serverDispatchCred{
+		ready: make(chan int),
+	}
+}
+func (c *serverDispatchCred) ClientHandshake(ctx context.Context, addr string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return rawConn, nil, nil
+}
+func (c *serverDispatchCred) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	c.rawConn = rawConn
+	close(c.ready)
+	return nil, nil, credentials.ErrConnDispatched
+}
+func (c *serverDispatchCred) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{}
+}
+func (c *serverDispatchCred) GetRawConn() net.Conn {
+	<-c.ready
+	return c.rawConn
+}
+
+func TestServerCredsDispatch(t *testing.T) {
+	lis, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatalf("Failed to listen: %v", err)
+	}
+	cred := newServerDispatchCred()
+	s := grpc.NewServer(grpc.Creds(cred))
+	go s.Serve(lis)
+	defer s.Stop()
+
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(cred))
+	if err != nil {
+		t.Fatalf("Dial(%q) = %v", lis.Addr().String(), err)
+	}
+	defer cc.Close()
+
+	go func() {
+		tc := testpb.NewTestServiceClient(cc)
+		tc.EmptyCall(context.Background(), &testpb.Empty{}, grpc.FailFast(false))
+	}()
+
+	buf := make([]byte, 1, 1)
+	// Check rawConn is not closed.
+	if n, err := cred.GetRawConn().Read(buf); n <= 0 || err != nil {
```

Comment:
write might be better here (because it does not need to send an rpc in line#2396).